Web3/Classes/Core/Contract/EthereumContract.swift

```diff
@@ -0,0 +1,178 @@
+    }
+    
+    /// Adds a method object to list of stored methods. Generally this should be done automatically by Web3.
+    ///
+    /// - Parameter method: `ABIFunction` that can be called on this contract
+    public func add(method: SolidityFunction) {
+        methods[method.name] = method
+    }
+    
+    /// Invocation of a method with the provided name
+    /// For example: `MyContract['balanceOf']?(address).call() { ... }`
+    ///
+    /// - Parameter name: Name of function to call
+    public subscript(_ name: String) -> ((ABIConvertible...) -> SolidityInvocation)? {
+        return methods[name]?.invoke
+    }
+    
+    /// Deploys a new instance of this contract to the network
+    /// Example: contract.deploy(byteCode: byteCode, parameters: p1, p2)?.send(...) { ... }
+    ///
+    /// - Parameters:
+    ///   - byteCode: Compiled bytecode of the contract
+    ///   - parameters: Any input values for the constructor
+    /// - Returns: Invocation object that can be called with .send(...)
+    public func deploy(byteCode: EthereumData, parameters: ABIConvertible...) -> SolidityConstructorInvocation? {
+        return constructor?.invoke(byteCode: byteCode, parameters: parameters)
+    }
+    
+    public func deploy(byteCode: EthereumData) -> SolidityConstructorInvocation? {
+        return constructor?.invoke(byteCode: byteCode, parameters: [])
+    }
+}
+
+// MARK: - Call & Send
+
+extension EthereumContract {
+    
+    /// Returns data by calling a constant function on the contract
+    ///
+    /// - Parameters:
+    ///   - data: EthereumData object representing the method called
+    ///   - outputs: Expected return values
+    ///   - completion: Completion handler
+    public func call(_ call: EthereumCall, outputs: [SolidityFunctionParameter], block: EthereumQuantityTag = .latest, completion: @escaping ([String: Any]?, Error?) -> Void) {
+        eth.call(call: call, block: block) { response in
+            switch response.status {
+            case .success(let data):
+                do {
+                    let dictionary = try ABI.decodeParameters(outputs, from: data.hex())
+                    completion(dictionary, nil)
+                } catch {
+                    completion(nil, error)
+                }
+            case .failure(let error):
+                completion(nil, error)
+            }
+        }
+    }
+    
+    /// Modifies the contract's data by sending a transaction
+    ///
+    /// - Parameters:
+    ///   - data: Encoded EthereumData for the methods called
+    ///   - from: EthereumAddress to send from
+    ///   - value: Amount of ETH to send, if applicable
+    ///   - gas: Maximum gas allowed for the transaction
+    ///   - gasPrice: Amount of wei to spend per unit of gas
+    ///   - completion: completion handler. Either the transaction's hash or an error.
+    public func send(_ transaction: EthereumTransaction, completion: @escaping (EthereumData?, Error?) -> Void) {
```

Comment:
Could we also add another convenient function to send raw transactions? Like `sendRawTransaction`?

Reply:
I think this one also applies to the protocol `SolidityFunctionHandler`

Reply:
So how would this one work? You'd have to sign the transaction somehow, so are you suggesting that it would take a private key, or thinking that it would just forward to the provider? These methods are mostly proxies for instances of SolidityFunction, so it could be a little strange in that case.

Reply:
Forget it. It was a fallacy on my side...